### PR TITLE
Linksharing with object map

### PR DESCRIPTION
**Pull request by boshevski**

### CODE REVIEW

File: linksharing/handler.go [hunk 1]
```diff
@@ -60,10 +70,16 @@ func NewHandler(log *zap.Logger, config HandlerConfig) (*Handler, error) {
 		return nil, err
 	}
 
+	mapper, err := objectmap.NewIPDB("./static/maxMindDB.mmdb")
```
Comment (grafael) [inline linksharing/handler.go:73]: It is preferable to load the `maxmind` database file path from config. Maxmind databases are updated very often.

Comment (grafael) [reply linksharing/handler.go:73]: Or you can keep this and enforce to always use the same maxmind database name.

File: go.mod [hunk 2]
```diff
@@ -6,10 +6,13 @@ require (
 	github.com/btcsuite/btcutil v1.0.2 // indirect
 	github.com/calebcase/tmpfile v1.0.2 // indirect
 	github.com/gogo/protobuf v1.3.1 // indirect
+	github.com/oschwald/geoip2-golang v1.4.0
 	github.com/oschwald/maxminddb-golang v1.7.0
 	github.com/spacemonkeygo/monkit/v3 v3.0.7-0.20200515175308-072401d8c752
 	github.com/spf13/cobra v1.0.0
+	github.com/storj/storjmap v0.0.0-20200616005132-b4224cbbe795
```
Comment (grafael) [inline go.mod:13]: You should not use this: `github.com/storj/storjmap`

File: go.sum [hunk 3]
```diff
@@ -223,6 +228,8 @@ github.com/spf13/pflag v1.0.5 h1:iy+VFUOCP1a+8yFto/drg2CJ5u0yRoB7fZw3DKv/JXA=
 github.com/spf13/pflag v1.0.5/go.mod h1:McXfInJRrz4CZXVZOBLb0bTZqETkiAhM9Iw0y3An2Bg=
 github.com/spf13/viper v1.4.0 h1:yXHLWeravcrgGyFSyCgdYpXQ9dR9c/WED3pg1RhxqEU=
 github.com/spf13/viper v1.4.0/go.mod h1:PTJ7Z/lr49W6bUbkmS1V3by4uWynFiR9p7+dSq/yZzE=
+github.com/storj/storjmap v0.0.0-20200616005132-b4224cbbe795 h1:9tnWoz8yUOxJz0DD7M8f1OzGLFhFtuCILxpZauQDBgY=
+github.com/storj/storjmap v0.0.0-20200616005132-b4224cbbe795/go.mod h1:wy9Jm/8hQccuvnNbT+MGv1kgE2V9h0w/rHD5psLTon0=
```
Comment (grafael) [inline go.sum:232]: Same as above.

**Review by grafael — CHANGES_REQUESTED** [review]
There are few minor changes regarding the `go.mod` . Also, check the possibility to use a configuration file for maxmind database.

**Review by grafael — CHANGES_REQUESTED** [review]
Remove the `maxMindDB.mmdb` binary file.  It should not be part of the PR (we cannot distribute this file). 
To help avoid pushing maxmind database, add these line on `.gitignore`
```bash
# avoid pushing maxmind databases
*.mmdb
```
Also double check the tests, they are failing.

File: linksharing/handler.go [hunk 4]
```diff
@@ -136,14 +152,52 @@ func (handler *Handler) serveHTTP(w http.ResponseWriter, r *http.Request) (err e
...
+			if err != nil {
+				handler.log.Error("failed to get IP info", zap.Error(err))
+				continue
+			}
+
+			ipInfos = append(ipInfos, *info)
+		}
+
+		var locations []Location
+		for _, info := range ipInfos {
```
Comment (mniewrzal) [inline linksharing/handler.go:179]: no need to have 3 loops here, try something like this
```
		var locations []Location
		for _, ip := range ipBytes {
			info, err := handler.mapper.GetIPInfos(string(ip))
			if err != nil {
				handler.log.Error("failed to get IP info", zap.Error(err))
				continue
			}

			location := Location{
				Latitude:  info.Location.Latitude,
				Longitude: info.Location.Longitude,
			}

			locations = append(locations, location)
		}
```

File: linksharing/handler.go [hunk 5]
```diff
@@ -136,14 +152,52 @@ func (handler *Handler) serveHTTP(w http.ResponseWriter, r *http.Request) (err e
...
+			Pieces int64
 		}
-		input.Name = bucket + "/" + o.Key
+		input.Name = o.Key
 		input.Size = memory.Size(o.System.ContentLength).Base10String()
+		input.Locations = locations
+		input.Pieces = int64(len(locations))
 
-		return handler.templates.ExecuteTemplate(w, "single-object.html", input)
+		err = handler.templates.ExecuteTemplate(w, "single-object.html", input)
```
Comment (mniewrzal) [inline linksharing/handler.go:199]: why creating `err`? return directly `handler.templates.ExecuteTemplate()`

**Review by mniewrzal — CHANGES_REQUESTED** [review]
I cannot see map in my local environment
Screenshot from 2020-10-15 10-10-29

Other than that + @grafael comments looks nice :-)

File: linksharing/handler.go (hunk 1 above)
Comment (VitaliiShpital) [reply linksharing/handler.go]: done

File: go.mod (hunk 2 above)
Comment (VitaliiShpital) [reply go.mod]: done

File: linksharing/handler.go (hunk 4 above)
Comment (VitaliiShpital) [reply linksharing/handler.go]: done

File: linksharing/handler.go (hunk 5 above)
Comment (VitaliiShpital) [reply linksharing/handler.go]: is was done just for debugging. Removed it

File: objectmap/mapper_test.go [hunk 6]
```diff
@@ -4,46 +4,12 @@
...
-	"errors"
-	"net"
 	"testing"
 
 	"github.com/stretchr/testify/require"
 )
 
-type MockReader struct{}
-
-func (mr *MockReader) Lookup(ip net.IP, result interface{}) error {
```
Comment (grafael) [inline objectmap/mapper_test.go:16]: Why are you removing these from mapper_test?

**Review by grafael — COMMENTED** [review]
There is just one comment regarding the `mapper_test`

**Review by grafael — DISMISSED** [review]

File: main.go [hunk 7]
```diff
@@ -82,7 +85,14 @@ func cmdRun(cmd *cobra.Command, args []string) (err error) {
 		}
 	}
 
-	handler, err := linksharing.NewHandler(log, linksharing.HandlerConfig{URLBase: runCfg.PublicURL})
+	reader, err := maxminddb.Open(runCfg.GeoLocationDB)
```
Comment (mniewrzal) [inline main.go:88]: it looks that reader needs to be closed

File: main.go [hunk 8]
```diff
@@ -82,7 +85,14 @@ func cmdRun(cmd *cobra.Command, args []string) (err error) {
...
+	if err != nil {
+		return err
+	}
+
+	mapper := objectmap.NewIPDB(reader)
```
Comment (mniewrzal) [inline main.go:93]: it looks that mapper needs to be closed

**Review by mniewrzal — APPROVED** [review]
@VitaliiShpital @boshevski Please, in future, try to split such large changes into smaller PRs :P

**Review by VitaliiShpital — APPROVED** [review]

**Review by grafael — APPROVED** [review]

**Review by crawter — APPROVED** [review]

**Pull request merged**